### ISSUE TITLE
pyRevit Loader Error Revit 2019 pyRevit v4.7.4

### ISSUE THREAD

**Issue by jagostinhoCT** [eirannejad/pyRevit#760]
getting this error when starting Revit 2019.

![pyRevit Loaded Error](https://user-images.githubusercontent.com/43812310/69745669-92fc3900-113a-11ea-8dc3-52b109bb9dbd.png)

**Comment by Kaankrby** [comment]
![Ek Açıklama 2019-11-28 233411](https://user-images.githubusercontent.com/10130515/69830837-ab9e4900-1237-11ea-98d6-ff202bf638e9.png)
Similar here.

> Not sure if we tried this before or not but open a command prompt and run pyrevit caches clear --all to clear all the exiting caches.

> Also look into your pyRevit config file (%APPDATA%/pyRevit/pyRevit_config.ini) and make sure bincache = true



tried this and nothing changes.

**Comment by eirannejad** [comment]
Would you please confirm that `Xceed.Wpf.AvalonDock.dll` exists in the Revit directory?

**Comment by Kaankrby** [comment]
I tried these;

> Not sure if we tried this before or not but open a command prompt and run pyrevit caches clear --all to clear all the exiting caches. Also look into your pyRevit config file (%APPDATA%/pyRevit/pyRevit_config.ini) and make sure bincache = true

> Seems like you are missing the pyRevitLabs.TargetApps.Revit dll. Make sure antivirus doesn't accidentally delete the file and is available under the bin/ directory. Install multiple times if needed to test.

> Okay. Which CLI version are you using? run pyrevit --version
> Add the installed clone using
> pyrevit clones add master "C:\Users\arthur.banks\AppData\Roaming\pyRevit-Master"
> Then run the pyrevit env again and verify master is listed under ==> Registered Clones (deployed from archive)
> Now reattach the clone to the installed Revits using
> pyrevit attach master 277 --installed

Would you please confirm that Xceed.Wpf.AvalonDock.dll exists in the Revit directory?

Yes it exist
![image](https://user-images.githubusercontent.com/10130515/69889869-51b78500-1304-11ea-8b0e-1978dd96cd77.png)

**Comment by jagostinhoCT** [comment]
![image](https://user-images.githubusercontent.com/43812310/69957310-5180da80-14fa-11ea-973d-c29d1c70bd62.png)
Yes. it is present.

**Comment by deddolo** [comment]
I have recently deployed 4.7.4 to my company computers and one user is having this same issue.

**Comment by Kaankrby** [comment]
> I have recently deployed 4.7.4 to my company computers and one user is having this same issue.
> 
> EDIT: Clearing the cache with `pyrevit caches clear 2019` seems to have solved the issue.

No this dosent solved the problem.

**Comment by Kaankrby** [comment]
i made a clean install deleted all of the files but still no change. 

İ don't know it's normal or not but, after clean install, the config file dosnt contain these lines

> 

> [environment]
> clones = {"master":"C:\\Users\\ITopya\\AppData\\Roaming\\pyRevit-Master"}
> 
> [core]
> checkupdates = false
> autoupdate = false
> verbose = true
> debug = false
> filelogging = false
> startuplogtimeout = 10
> userextensions = []
> compilecsharp = true
> compilevb = true
> loadbeta = false
> rocketmode = true
> bincache = true
> usercanupdate = true
> usercanextend = true
> usercanconfig = true
> 
> [usagelogging]
> active = false
> logfilepath = ""
> logserverurl = ""
> 
> [pyRevitTags.extension]
> disabled = true
> private_repo = true
> username = ""
> password = ""
> 
> [pyRevitTools.extension]
> disabled = false
> private_repo = true
> username = ""
> password = ""
> 
> [pyRevitTutor.extension]
> disabled = true
> private_repo = true
> username = ""
> password = ""

> 
only this one 


> [environment]
> clones = {"master":"C:\\Users\\ITopya\\AppData\\Roaming\\pyRevit-Master"}

and i added 

> bincache = true

**Comment by Kaankrby** [comment]
Copied these lines to 4.7.4 config file. nothing changed

**Comment by eirannejad** [comment]
@Kaankrby There rest of the config lines are added when pyRevit loads for the first time.

**Comment by Kaankrby** [comment]
> @Kaankrby There rest of the config lines are added when pyRevit loads for the first time.

I don't have any coding skill :) i try what i can.
Still it's not working, using 4,6

**Comment by Kaankrby** [comment]
is there any update

**Comment by eirannejad** [comment]
Try the new 4.7.5 and see if this has been fixed

**Comment by Kaankrby** [comment]
Sorry but still the same
![Ek Açıklama 2020-03-18 170137](https://user-images.githubusercontent.com/10130515/76968750-74fd5300-693a-11ea-8b4e-bdee9708f008.png)

**Comment by eirannejad** [comment]
This might be a conflict with another addon. Try disabling the other addons using [this tool](https://www.rhino3d.com/inside/revit/beta/reference/toubleshooting#search-for-conflicting-plugins) and see if pyRevit loads by itself.

**Comment by Kaankrby** [comment]
tried but nothing changes. 
In turkey, i try in 3 different office and may be 10 pc. 
Always the same. 

Can be a regional problem. 
Some of the pc languege was English too. 

wil try with a clean install rvt20

**Comment by eirannejad** [comment]
Okay would you mind:
- Close Revit
- Open `/pyrevitlib/pyrevit/loader/sessionmgr.py` in your pyRevit install path (default is `%APPDATA%/pyRevit-Master`
- On line 67 Change the `_setup_output` function to this:

```python
def _setup_output():
    # create output window and assign handle
    with open(op.join(USER_DESKTOP, 'debug.log'), 'w') as f:
        f.writelines(
            "runtime is loaded as: {}"
            "runtime.types is loaded as: {}"
            "runtime.types.ScriptConsole is loaded as: {}\n".format(
                runtime,
                runtime.types,
                runtime.types.ScriptConsole
            ))
        try:
            out_window = runtime.types.ScriptConsole()
        except Exception as ex:
            f.writelines(str(ex))

    out_window = runtime.types.ScriptConsole()
    runtime_info = sessioninfo.get_runtime_info()
    out_window.AppVersion = '{}:{}:{}'.format(
        runtime_info.pyrevit_version,
        runtime_info.engine_version,
        runtime_info.host_version
        )

    # create output stream and set stdout to it
    # we're not opening the output window here.
    # The output stream will open the window if anything is being printed.
    outstr = runtime.types.ScriptIO(out_window)
    sys.stdout = outstr
    # sys.stderr = outstr
    stdout_hndlr = logger.get_stdout_hndlr()
    stdout_hndlr.stream = outstr

    return out_window
```

- Then open a command prompt and run `pyrevit caches clear --all`
- Run Revit again
- This time pyRevit should create a `debug.log` file on your desktop.

Send me that file please

**Comment by Kaankrby** [comment]
![sdds](https://user-images.githubusercontent.com/10130515/77482065-223e0280-6e36-11ea-86b4-5937cc8790d7.JPG)

code dosn't work.


added

` import os `

and changed 

` with open(op.join(op.join(op.join(os.environ['USERPROFILE']), 'Desktop'), 'debug.log'), 'w') as f:`

[debug.log](https://github.com/eirannejad/pyRevit/files/4377897/debug.log)

*i am not an expert, writing what i am doing.

**Comment by eirannejad** [comment]
Ugh sorry. Modify line 17 at the top as well:

`from pyrevit import EXEC_PARAMS, HOST_APP, USER_DESKTOP`

**Comment by eirannejad** [comment]
Okay let me see if I can offer a few more changes to better debug the issue

**Comment by Kaankrby** [comment]
> Ugh sorry. Modify line 17 at the top as well:
> 
> `from pyrevit import EXEC_PARAMS, HOST_APP, USER_DESKTOP`
> 
> This should work now `open(op.join(USER_DESKTOP, 'debug.log')`

made these. 

[debug.log](https://github.com/eirannejad/pyRevit/files/4377919/debug.log)

**Comment by eirannejad** [comment]
Okay. One more try. Replace the previous function with this:

```python
def _setup_output():
    # create output window and assign handle
    try:
        out_window = runtime.types.ScriptConsole()
    except Exception as ex:
        with open(op.join(USER_DESKTOP, 'debug.log'), 'w') as f:
            f.writelines(ex.clsException.Message)
            f.writelines(ex.clsException.Source)
            f.writelines(ex.clsException.StackTrace)
            f.writelines(ex.clsException.TargetSite.ToString())
    # runtime_info = sessioninfo.get_runtime_info()
    # out_window.AppVersion = '{}:{}:{}'.format(
    #     runtime_info.pyrevit_version,
    #     runtime_info.engine_version,
    #     runtime_info.host_version
    #     )

    # # create output stream and set stdout to it
    # # we're not opening the output window here.
    # # The output stream will open the window if anything is being printed.
    # outstr = runtime.types.ScriptIO(out_window)
    # sys.stdout = outstr
    # # sys.stderr = outstr
    # stdout_hndlr = logger.get_stdout_hndlr()
    # stdout_hndlr.stream = outstr

    # return out_window

```

Then run again and take a screenshot of the error message that pops up

**Comment by Kaankrby** [comment]
![Ekran Alıntısı](https://user-images.githubusercontent.com/10130515/77486063-c9736780-6e3f-11ea-86c5-8a92743ba3e7.JPG)

**Comment by eirannejad** [comment]
Ugh. This went somewhere else completely. Hmmmmm. Okay give me time to think a bit

**Comment by Kaankrby** [comment]
> i think this will help. 
> 
> windows has a region setting and language setting. 
> 
> i changed both of them to U.S. 
> Now my pc thinks i live in U.S and my pc language English too. 
> And the result. 

![Annotation 2020-03-25 1458312](https://user-images.githubusercontent.com/10130515/77534790-5d7a1900-6eaa-11ea-9b2b-10d8ec1fcff8.jpg)
![Annotation 2020-03-25 145831](https://user-images.githubusercontent.com/10130515/77534795-5f43dc80-6eaa-11ea-9b9b-610a8e440bf9.jpg)

> Yes its working. 
> Any problem. 
> 
> But when i change region to Turkey. Still language is English. 
> 

![reg2](https://user-images.githubusercontent.com/10130515/77534870-83072280-6eaa-11ea-8a34-3d789672acce.jpg)

![Ekran Alıntısı](https://user-images.githubusercontent.com/10130515/77534906-9ca86a00-6eaa-11ea-8f1f-1ca905582e7b.JPG)


> Same issue. 
> 
> and the last one.

![regg](https://user-images.githubusercontent.com/10130515/77535108-13456780-6eab-11ea-952b-0855bcf6f730.jpg)

> regional settings US
> 
> region TURKEY
> 
> it works.

![ok](https://user-images.githubusercontent.com/10130515/77535156-2d7f4580-6eab-11ea-9167-7470a4f9d9e3.jpg)


> This option changes Date format and number format. 
> and i find; 
> 


> turkey number:

![TURKEY](https://user-images.githubusercontent.com/10130515/77535788-40dee080-6eac-11ea-8ed6-d8ecdd3fda75.jpg)

> usa number: 


![usa](https://user-images.githubusercontent.com/10130515/77535796-46d4c180-6eac-11ea-873e-3e13d79649b8.jpg)

**Comment by Kaankrby** [comment]
this wasnt a problem for 4,6
staring with beta of 4,7 this problem occurs.

**Comment by Kaankrby** [comment]
![english eurpoe](https://user-images.githubusercontent.com/10130515/77540352-fa8d7f80-6eb3-11ea-8d4c-79aedffe61cb.jpg)



i think the problem dosnt related directly to number decimal symbol. 
Because its work for europe too. 
And europe is same Turkey

**Comment by Kaankrby** [comment]
when all the options are correct, i start revit and pyrevit loads well. 
then i change the local to TURKEY.

i reload pyrevit its works well

**Comment by eirannejad** [comment]
Okay great! I can replicate the issue now. Let me dig into it and figure out why this happens

**Comment by eirannejad** [comment]
Solved. In C# the method `String.ToUpper()` uses the active locale so `pyRevit` was being capitalized as `PYREVÍT` in Turkish

- Cancel out all the previous code changes
- Open `pyrevitlib/pyrevit/runtime/EnvVariables.cs`
- Change the `.ToUpper()` call to `.ToUpperInvariant()` on lines 11 and 21

```csharp
namespace PyRevitLabs.PyRevit.Runtime {
    public static class DomainStorageKeys {
        public static string keyPrefix = PyRevitLabsConsts.ProductName.ToUpperInvariant();

        public static string EnvVarsDictKey = keyPrefix + "EnvVarsDict";
        public static string EnginesDictKey = keyPrefix + "CachedEngines";
        public static string IronPythonEngineDefaultStreamCfgKey = keyPrefix + "CachedEngineDefaultStreamCfg";
        public static string OutputWindowsDictKey = keyPrefix + "OutputWindowsDict";
    }

    public static class EnvDictionaryKeys
    {
        public static string keyPrefix = PyRevitLabsConsts.ProductName.ToUpperInvariant();

...
```

I'll make changes to the code and will publish this fix with the next 4.7.6 release soon
